### PR TITLE
Call .decode() only for binary data

### PR DESCRIPTION
**Pull request by frenzymadness**
## Summary of changes

I believe that the `__init__` method might be simplified in this way. In my opinion, it makes sense to call `decode()` method only on bytes/bytearray objects in both major Pythons.

In Python 2, `bytes` is builtin alias for `str` so `decode()` call does not happen for unicode.

Do you know about any corner case this simplified code is not aware of?

### Pull Request Checklist
- [ ] Changes have tests — this class is hopefully tested well enough
- [ ] Authors have been added to [AUTHORS.md](https://github.com/dateutil/dateutil/blob/master/AUTHORS.md) — not necessary, nothing really important contributed
- [ ] News fragment added in changelog.d. See [CONTRIBUTING.md](https://github.com/dateutil/dateutil/blob/master/CONTRIBUTING.md#changelog) for details — not important user-facing change

### CODE REVIEW

File: dateutil/parser/_parser.py [hunk 1]
```diff
@@ -60,14 +60,8 @@ class _timelex(object):
...
     def __init__(self, instream):
-        if six.PY2:
-            # In Python 2, we can't duck type properly because unicode has
-            # a 'decode' function, and we'd be double-decoding
-            if isinstance(instream, (bytes, bytearray)):
-                instream = instream.decode()
-        else:
-            if getattr(instream, 'decode', None) is not None:
-                instream = instream.decode()
+        if isinstance(instream, (bytes, bytearray)):
```
Comment (vstinner) [inline dateutil/parser/_parser.py:63]: I suggest to also accept memoryview. For exampel, asyncio uses:

```
        if not isinstance(data, (bytes, bytearray, memoryview)):
            raise TypeError(f'data argument must be a bytes-like object, '
                            f'not {type(data).__name__!r}')
```

Comment (vstinner) [reply dateutil/parser/_parser.py:63]: I don't know if dateutil still supports Python 2. In that case, you may need to only use memoryview on Python 3.

Comment (frenzymadness) [reply dateutil/parser/_parser.py]: Implemented, thank you!

File: dateutil/parser/_parser.py [hunk 2]
```diff
@@ -60,14 +60,8 @@ class _timelex(object):
...
     def __init__(self, instream):
-        if six.PY2:
-            # In Python 2, we can't duck type properly because unicode has
-            # a 'decode' function, and we'd be double-decoding
-            if isinstance(instream, (bytes, bytearray)):
-                instream = instream.decode()
-        else:
-            if getattr(instream, 'decode', None) is not None:
-                instream = instream.decode()
+        if isinstance(instream, (bytes, bytearray)) or six.PY3 and isinstance(instream, memoryview):
```
Comment (dmach) [inline dateutil/parser/_parser.py:63]: Wouldn't be better to do duck-typing and test for the presence of the decode() method?

```
if hasattr(instream, "decode"):
    instream = instream.decode()
```

Comment (mariocj89) [reply dateutil/parser/_parser.py:63]: If you pass in Python 2 a unicode string it will still attempt to decode it by encoding it first. And if that string has a non-ascii char, it'd fail to encode.

Comment (mariocj89) [inline dateutil/parser/_parser.py:63]: I think that what @vstinner meant with supporting memoryview was not this. This will fail when a memoryview is passed, as it does not have a `decode` method AFAIK.

If you want to add support for memoryview, please add tests for it. Alternatively, feel free to just remove `or six.PY3 and isinstance(instream, memoryview)` from this PR.

**Review by mariocj89 — CHANGES_REQUESTED** [review]

**Comment by frenzymadness** [comment]
I must admit that I kinda lost the context of this change after the eleven months. Give me a while or, if you know the solution, feel free to open a new PR based on this one and finish the fix.

**Comment by mariocj89** [comment]
> Give me a while or, if you know the solution, feel free to open a new PR based on this one and finish the fix.

Sure, I'll push a commit with the additions.

File: dateutil/parser/_parser.py (hunk 2 above)
Comment (mariocj89) [inline dateutil/parser/_parser.py:63]: At the same time, I see that commit 8e786465437b51123ab7644b8ce94b51d947c5b8 says "Add support for bytesarray and other bytes-like", not sure if there is any other case we are missing here than `bytes`, `bytearray` and `unicode`.

Let's see if the test suite catches anything.

**Review by mariocj89 — APPROVED** [review]
LGTM, but would be good to get another review as I rewrote your commit.

**Comment by frenzymadness** [comment]
@vstinner could you please take a look?

**Review by vstinner — APPROVED** [review]
LGTM.

Ignore my remark about memoryview, memoryview objects have no decode() method.
